src/uucore/src/lib/mods/display.rs

```diff
@@ -0,0 +1,517 @@
+// for backward compatibility reasons. Otherwise we'd use a blanket impl.
+impl Quotable for Cow<'_, str> {
+    fn quote(&self) -> Quoted<'_> {
+        // I suspect there's a better way to do this but I haven't found one
+        Quoted::new(<Cow<'_, str> as Borrow<str>>::borrow(self).as_ref())
+    }
+}
+
+/// A wrapper around [`OsStr`] for printing paths with quoting and escaping applied.
+#[derive(Debug, Copy, Clone)]
+pub struct Quoted<'a> {
+    text: &'a OsStr,
+    force_quote: bool,
+    escape_control: bool,
+}
+
+impl<'a> Quoted<'a> {
+    fn new(text: &'a OsStr) -> Self {
+        Quoted {
+            text,
+            force_quote: true,
+            escape_control: true,
+        }
+    }
+
+    /// Add quotes even if not strictly necessary. `true` by default.
+    pub fn force_quote(mut self, force: bool) -> Self {
+        self.force_quote = force;
+        self
+    }
+
+    /// Escape control characters. `true` by default.
+    pub fn escape_control(mut self, escape: bool) -> Self {
+        self.escape_control = escape;
+        self
+    }
+}
+
+impl Display for Quoted<'_> {
+    #[cfg(any(unix, target_os = "wasi"))]
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        /// Characters with special meaning outside quotes.
+        // https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_02
+        // % seems obscure enough to ignore, it's for job control only.
+        // {} were used in a version elsewhere but seem unnecessary, GNU doesn't escape them.
+        // ! is a common extension.
+        const SPECIAL_SHELL_CHARS: &[u8] = b"|&;<>()$`\\\"'*?[]=! \t\n";
+        /// Characters with a special meaning at the beginning of a name.
+        const SPECIAL_SHELL_CHARS_START: &[u8] = b"~#";
+
+        let text = self.text.as_bytes();
+
+        let mut is_single_safe = true;
+        let mut is_double_safe = true;
+        let mut requires_quote = self.force_quote;
+
+        if let Some(first) = text.get(0) {
+            if SPECIAL_SHELL_CHARS_START.contains(first) {
+                requires_quote = true;
+            }
+        } else {
+            // Empty string
+            requires_quote = true;
+        }
+
+        for &ch in text {
+            match ch {
+                ch if self.escape_control && ch.is_ascii_control() => {
+                    return write_escaped(f, text, self.escape_control)
+                }
+                b'\'' => is_single_safe = false,
+                // Unsafe characters according to:
+                // https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_02_03
+                b'"' | b'`' | b'$' | b'\\' => is_double_safe = false,
+                _ => (),
+            }
+            if !requires_quote && SPECIAL_SHELL_CHARS.contains(&ch) {
+                requires_quote = true;
+            }
+        }
+        let text = match from_utf8(text) {
```

Comment:
I think we can move the conversion to a `&str` in front of the `for` loop above.